crates/noirc_frontend/src/lexer/token.rs

```diff
@@ -401,9 +403,10 @@ impl AsRef<str> for Attribute {
     }
 }
 
+/// All keywords in Noir.
+/// Note that `self` is not present - it is a contextual keyword rather than a true one as it is
+/// only special within `impl`s. Otherwise `self` functions as a normal identifier.
 #[derive(PartialEq, Eq, Hash, Debug, Copy, Clone, PartialOrd, Ord)]
```

Comment:
```suggestion
#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone, PartialOrd, Ord)]
#[cfg_attr(test, derive(strum_macros::EnumIter))]
```

We need to re-add the conditional derivation of this trait here. Looks like it got lost in the reorganisation.

Reply:
@jfecher does it mean that users can make function parameters named `self` and do things like `let self = ...` ?

Reply:
That is the case, yes. This isn't quite how it works in rust, but does follow other languages like python and lua.

Reply:
That sounds like it could cause confusion, since we want to be the ones who dictate where self can be placed.

Its not blocking for this PR as its about documenting existing behaviour.

Great work btw

Reply:
> That sounds like it could cause confusion, since we want to be the ones who dictate where self can be placed.

It seems fine to me to allow users to refer to `self` as if it were a normal identifier outside of an impl. Similar to contextual keywords in other languages. For example in rust if you want to pass self to a lambda, you can't refer to it as self, so you have to make a new name for it like `this`. We can also still dictate where `self` has special behavior with this scheme.

Reply:
using `this` sounds good to me, I'd prefer the more conservative approach and say that all keywords that the language has special meanings for cannot be used as an identifier no matter the context.